// zod-to-schema/dist/index.mjs
// src/zodToMongoose.ts
import mongoose, {
  Schema
} from "mongoose";
import {
  ZodArray,
  ZodBoolean,
  ZodDate,
  ZodDefault,
  ZodEnum,
  ZodLiteral,
  ZodNullable,
  ZodNumber,
  ZodObject,
  ZodOptional,
  ZodString
} from "zod";
function zodToMongoose(zodSchema, relationshipMappings) {
  const mongooseSchemaDefinition = {};
  const shape = zodSchema.shape;
  for (const [key, schema] of Object.entries(shape)) {
    const mongooseField = {};
    let currentSchema = schema;
    let isOptional = false;
    let hasDefault = false;
    let defaultValue;
    while (currentSchema instanceof ZodOptional || currentSchema instanceof ZodDefault || currentSchema instanceof ZodNullable) {
      if (currentSchema instanceof ZodOptional) {
        isOptional = true;
        currentSchema = currentSchema._def.innerType;
      }
      if (currentSchema instanceof ZodDefault) {
        hasDefault = true;
        defaultValue = currentSchema._def.defaultValue();
        currentSchema = currentSchema._def.innerType;
      }
      if (currentSchema instanceof ZodNullable) {
        currentSchema = currentSchema._def.innerType;
      }
    }
    if (currentSchema instanceof ZodArray) {
      const arraySchema = currentSchema;
      const itemType = arraySchema._def.type;
      if (relationshipMappings && relationshipMappings[key]) {
        mongooseField.type = [mongoose.Schema.Types.ObjectId];
        mongooseField.ref = relationshipMappings[key];
      } else {
        if (itemType instanceof ZodString) {
          mongooseField.type = [String];
        } else if (itemType instanceof ZodNumber) {
          mongooseField.type = [Number];
        } else if (itemType instanceof ZodBoolean) {
          mongooseField.type = [Boolean];
        } else if (itemType instanceof ZodObject) {
          mongooseField.type = [mongoose.Schema.Types.Mixed];
        } else {
          mongooseField.type = [mongoose.Schema.Types.Mixed];
        }
      }
    } else if (relationshipMappings && relationshipMappings[key]) {
      mongooseField.type = mongoose.Schema.Types.ObjectId;
      mongooseField.ref = relationshipMappings[key];
    } else {
      switch (currentSchema.constructor) {
        case ZodString:
          mongooseField.type = String;
          const stringSchema = currentSchema;
          stringSchema._def.checks.forEach((check) => {
            switch (check.kind) {
              case "min":
                mongooseField.minlength = check.value;
                break;
              case "max":
                mongooseField.maxlength = check.value;
                break;
              case "email":
                mongooseField.match = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
                break;
              case "regex":
                mongooseField.match = check.regex;
                break;
              case "includes":
                break;
              case "length":
                mongooseField.length = check.value;
                break;
              default:
                break;
            }
          });
          break;
        case ZodNumber:
          mongooseField.type = Number;
          const numberSchema = currentSchema;
          numberSchema._def.checks.forEach((check) => {
            switch (check.kind) {
              case "min":
                mongooseField.min = check.value;
                break;
              case "max":
                mongooseField.max = check.value;
                break;
              case "int":
                mongooseField.validate = {
                  validator: Number.isInteger,
                  message: "{VALUE} is not an integer value"
                };
                break;
              default:
                break;
            }
          });
          break;
        case ZodBoolean:
          mongooseField.type = Boolean;
          break;
        case ZodDate:
          mongooseField.type = Date;
          break;
        case ZodEnum:
          const enumSchema = currentSchema;
          mongooseField.type = String;
          mongooseField.enum = enumSchema.options;
          break;
        case ZodLiteral:
          const literalSchema = currentSchema;
          switch (typeof literalSchema.value) {
            case "string":
              mongooseField.type = String;
              break;
            case "number":
              mongooseField.type = Number;
              break;
            case "boolean":
              mongooseField.type = Boolean;
              break;
            default:
              mongooseField.type = mongoose.Schema.Types.Mixed;
          }
          mongooseField.enum = [literalSchema.value];
          break;
        case ZodObject:
          mongooseField.type = mongoose.Schema.Types.Mixed;
          break;
        default:
          mongooseField.type = mongoose.Schema.Types.Mixed;
      }
    }
    if (hasDefault) {
      mongooseField.default = defaultValue;
    }
    mongooseField.required = !isOptional;
    mongooseSchemaDefinition[key] = mongooseField;
  }
  return mongooseSchemaDefinition;
}
function createMongooseModel(modelName, zodSchema, relationshipMappings) {
  if (mongoose.models[modelName]) {
    return mongoose.models[modelName];
  }
  const mongooseSchemaDefinition = zodToMongoose(
    zodSchema,
    relationshipMappings
  );
  const schema = new Schema(mongooseSchemaDefinition, { timestamps: true });
  return mongoose.model(modelName, schema);
}
function zodToObject(zodSchema, data, options) {
  const schemaToUse = options?.partial && zodSchema instanceof ZodObject ? zodSchema.partial() : zodSchema;
  const parseResult = schemaToUse.safeParse(data);
  if (!parseResult.success) {
    const errorMessages = parseResult.error.errors.map((err) => `${err.path.join(".")} - ${err.message}`).join(", ");
    throw new Error(`Zod validation error: ${errorMessages}`);
  }
  return parseResult.data;
}

// src/zodToPrisma.ts
import {
  ZodArray as ZodArray2,
  ZodBoolean as ZodBoolean2,
  ZodDate as ZodDate2,
  ZodDefault as ZodDefault2,
  ZodEffects,
  ZodEnum as ZodEnum2,
  ZodLiteral as ZodLiteral2,
  ZodNativeEnum,
  ZodNullable as ZodNullable2,
  ZodNumber as ZodNumber2,
  ZodObject as ZodObject2,
  ZodOptional as ZodOptional2,
  ZodString as ZodString2
} from "zod";
var enums = [];
function getEnumName(fieldName) {
  return `${capitalize(fieldName)}Enum`;
}
function capitalize(str) {
  if (!str) return str;
  return str.charAt(0).toUpperCase() + str.slice(1);
}
function camelCase(str) {
  if (!str) return str;
  return str.charAt(0).toLowerCase() + str.slice(1);
}
function zodTypeToPrisma(zodType, fieldName, schemaToModelName) {
  if (zodType instanceof ZodEffects) {
    return zodTypeToPrisma(zodType._def.schema, fieldName, schemaToModelName);
  }
  if (zodType instanceof ZodString2) {
    return "String";
  }
  if (zodType instanceof ZodNumber2) {
    if (zodType._def.checks?.some((check) => check.kind === "int")) {
      return "Int";
    }
    return "Float";
  }
  if (zodType instanceof ZodBoolean2) {
    return "Boolean";
  }
  if (zodType instanceof ZodDate2) {
    return "DateTime";
  }
  if (zodType instanceof ZodArray2) {
    const elementType = zodType.element;
    const relatedModelName = schemaToModelName.get(elementType);
    if (relatedModelName) {
      return `${relatedModelName}[]`;
    }
    const prismaElementType = zodTypeToPrisma(
      elementType,
      fieldName,
      schemaToModelName
    );
    return `${prismaElementType}[]`;
  }
  if (zodType instanceof ZodEnum2) {
    const enumName = getEnumName(fieldName);
    if (!enums.find((e) => e.name === enumName)) {
      enums.push({
        name: enumName,
        values: zodType._def.values
      });
    }
    return enumName;
  }
  if (zodType instanceof ZodNativeEnum) {
    const enumObj = zodType._def.values;
    const enumValues = Object.values(enumObj).filter(
      (v) => typeof v === "string"
    );
    const enumName = getEnumName(fieldName);
    if (!enums.find((e) => e.name === enumName)) {
      enums.push({
        name: enumName,
        values: enumValues
      });
    }
    return enumName;
  }
  if (zodType instanceof ZodLiteral2) {
    const literalValue = zodType._def.value;
    if (typeof literalValue === "string") {
      return `String @default("${literalValue}")`;
    }
    if (typeof literalValue === "number") {
      return `Int @default(${literalValue})`;
    }
    if (typeof literalValue === "boolean") {
      return `Boolean @default(${literalValue})`;
    }
    throw new Error(`Unsupported ZodLiteral type: ${typeof literalValue}`);
  }
  if (zodType instanceof ZodObject2) {
    const relatedModelName = schemaToModelName.get(zodType);
    if (relatedModelName) {
      return relatedModelName;
    }
    return "Json";
  }
  throw new Error(`Unsupported Zod type: ${zodType.constructor.name}`);
}
function extractPrismaFields(zodObject, schemaToModelName, modelName, modelNameToRelations, modelNameToSchema) {
  const shape = zodObject.shape;
  const fields = [];
  if (!shape.hasOwnProperty("id")) {
    fields.push({
      name: "id",
      type: "String",
      isRequired: true,
      isUnique: true,
      isId: true,
      isAutoIncrement: false,
      default: "uuid()"
    });
  }
  for (const [key, value] of Object.entries(shape)) {
    if (key === "id") {
      fields.push({
        name: "id",
        type: "String",
        isRequired: true,
        isId: true,
        isUnique: true,
        isAutoIncrement: false,
        default: "uuid()"
      });
      continue;
    }
    let isRequired = true;
    let isUnique = false;
    const isId = false;
    const isAutoIncrement = false;
    let defaultValue;
    let isEnum = false;
    let isRelation = false;
    let relationModel;
    let relationName;
    const relationFields = [];
    const references = [];
    let zodType = value;
    while (zodType instanceof ZodOptional2 || zodType instanceof ZodNullable2 || zodType instanceof ZodDefault2 || zodType instanceof ZodEffects) {
      if (zodType instanceof ZodOptional2 || zodType instanceof ZodNullable2) {
        isRequired = false;
        zodType = zodType.unwrap();
      } else if (zodType instanceof ZodDefault2) {
        const defaultVal = zodType._def.defaultValue();
        zodType = zodType._def.innerType;
        if (zodType instanceof ZodBoolean2) {
          defaultValue = defaultVal ? "true" : "false";
        } else if (zodType instanceof ZodString2) {
          defaultValue = `"${defaultVal}"`;
        } else if (zodType instanceof ZodNumber2) {
          defaultValue = `${defaultVal}`;
        } else if (zodType instanceof ZodDate2) {
          defaultValue = "now()";
        } else if (zodType instanceof ZodEnum2 || zodType instanceof ZodNativeEnum) {
          defaultValue = `${defaultVal}`;
        }
      } else if (zodType instanceof ZodEffects) {
        zodType = zodType._def.schema;
      }
    }
    if (key.toLowerCase().includes("email")) {
      isUnique = true;
    }
    if (zodType instanceof ZodEnum2 || zodType instanceof ZodNativeEnum) {
      isEnum = true;
    }
    if (zodType instanceof ZodObject2 || zodType instanceof ZodArray2) {
      const elementType = zodType instanceof ZodArray2 ? zodType.element : zodType;
      if (elementType instanceof ZodObject2) {
        const relatedModelName = schemaToModelName.get(elementType);
        if (relatedModelName) {
          isRelation = true;
          relationModel = relatedModelName;
          relationName = `${modelName}To${relatedModelName}`;
          const relatedRelations = modelNameToRelations.get(relatedModelName) || [];
          if (zodType instanceof ZodArray2) {
            relatedRelations.push({
              relationName,
              relatedModel: modelName,
              type: "many",
              fieldName: key
            });
          } else {
            relatedRelations.push({
              relationName,
              relatedModel: modelName,
              type: "one",
              fieldName: key
            });
          }
          modelNameToRelations.set(relatedModelName, relatedRelations);
        }
      }
    }
    if (key.endsWith("Id") && (zodType instanceof ZodString2 || zodType instanceof ZodNumber2)) {
      const relatedModelName = capitalize(key.slice(0, -2));
      if (modelNameToSchema.has(relatedModelName)) {
        fields.push({
          name: key,
          type: zodTypeToPrisma(zodType, key, schemaToModelName),
          isRequired,
          isUnique: true
          // Set to true for one-to-one relations
        });
        fields.push({
          name: key.slice(0, -2),
          type: relatedModelName,
          isRequired,
          isRelation: true,
          relationModel: relatedModelName,
          relationName: `${modelName}To${relatedModelName}`,
          relationFields: [key],
          references: ["id"]
        });
        const relatedRelations = modelNameToRelations.get(relatedModelName) || [];
        relatedRelations.push({
          relationName: `${modelName}To${relatedModelName}`,
          relatedModel: modelName,
          type: "one",
          fieldName: void 0
        });
        modelNameToRelations.set(relatedModelName, relatedRelations);
        continue;
      }
    }
    let prismaType;
    try {
      prismaType = zodTypeToPrisma(zodType, key, schemaToModelName);
    } catch (error) {
      console.error(`Error processing field "${key}": ${error.message}`);
      throw error;
    }
    fields.push({
      name: key,
      type: prismaType,
      isRequired,
      isUnique,
      isId,
      isAutoIncrement,
      default: defaultValue,
      isEnum,
      isRelation,
      relationModel,
      relationName,
      relationFields,
      references
    });
  }
  return fields;
}
function generatePrismaModel(modelName, zodSchema, schemaToModelName, modelNameToRelations, modelNameToSchema) {
  if (!(zodSchema instanceof ZodObject2)) {
    throw new Error(
      "Only ZodObject schemas are supported for Prisma model generation."
    );
  }
  const fields = extractPrismaFields(
    zodSchema,
    schemaToModelName,
    modelName,
    modelNameToRelations,
    modelNameToSchema
  );
  const prismaFields = [];
  fields.forEach((field) => {
    let line = `  ${field.name} ${field.type}`;
    if (!field.isRequired && !field.isRelation) {
      line += "?";
    }
    if (field.isId) line += " @id";
    if (field.isUnique) line += " @unique";
    if (field.isAutoIncrement) line += " @default(autoincrement())";
    if (field.default && !field.isEnum) {
      line += ` @default(${field.default})`;
    }
    if (field.isEnum && field.default) {
      line += ` @default(${field.default})`;
    }
    if (field.isRelation && field.relationModel) {
      if (field.relationFields?.length) {
        line += ` @relation("${field.relationName}", fields: [${field.relationFields.join(
          ", "
        )}], references: [${field.references?.join(", ")}])`;
      } else if (field.type.endsWith("[]")) {
        line += ` @relation("${field.relationName}")`;
      } else {
        line += ` @relation("${field.relationName}")`;
      }
    }
    prismaFields.push(line);
  });
  const relations = modelNameToRelations.get(modelName) || [];
  relations.forEach((rel) => {
    const fieldName = rel.fieldName || camelCase(rel.relatedModel) + (rel.type === "many" ? "s" : "");
    const fieldType = `${rel.relatedModel}${rel.type === "many" ? "[]" : "?"}`;
    const line = `  ${fieldName} ${fieldType} @relation("${rel.relationName}")`;
    prismaFields.push(line);
  });
  return `model ${modelName} {
${prismaFields.join("\n")}
}`;
}
function generatePrismaEnums() {
  return enums.map((enumObj) => {
    const values = enumObj.values.map((v) => `  ${v}`).join("\n");
    return `enum ${enumObj.name} {
${values}
}`;
  }).join("\n\n");
}
function zodToPrisma(models, schemaToModelName, modelNameToSchema) {
  const modelNameToRelations = /* @__PURE__ */ new Map();
  const prismaModels = models.map(
    ({ name, schema }) => generatePrismaModel(
      name,
      schema,
      schemaToModelName,
      modelNameToRelations,
      modelNameToSchema
    )
  ).join("\n\n");
  const prismaEnums = generatePrismaEnums();
  const prismaSchemaHeader = `
datasource db {
  provider = "postgresql"
  url      = env("POSTGRESQL_DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}
`;
  return `${prismaSchemaHeader}

${prismaEnums}

${prismaModels}`;
}
function handleRelations(modelString, schema) {
  return modelString;
}
export {
  createMongooseModel,
  generatePrismaEnums,
  generatePrismaModel,
  handleRelations,
  zodToMongoose,
  zodToObject,
  zodToPrisma
};
