// zod-to-schema/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  createMongooseModel: () => createMongooseModel,
  generatePrismaEnums: () => generatePrismaEnums,
  generatePrismaModel: () => generatePrismaModel,
  handleRelations: () => handleRelations,
  zodToMongoose: () => zodToMongoose,
  zodToObject: () => zodToObject,
  zodToPrisma: () => zodToPrisma
});
module.exports = __toCommonJS(src_exports);

// src/zodToMongoose.ts
var import_mongoose = __toESM(require("mongoose"));
var import_zod = require("zod");
function zodToMongoose(zodSchema, relationshipMappings) {
  const mongooseSchemaDefinition = {};
  const shape = zodSchema.shape;
  for (const [key, schema] of Object.entries(shape)) {
    const mongooseField = {};
    let currentSchema = schema;
    let isOptional = false;
    let hasDefault = false;
    let defaultValue;
    while (currentSchema instanceof import_zod.ZodOptional || currentSchema instanceof import_zod.ZodDefault || currentSchema instanceof import_zod.ZodNullable) {
      if (currentSchema instanceof import_zod.ZodOptional) {
        isOptional = true;
        currentSchema = currentSchema._def.innerType;
      }
      if (currentSchema instanceof import_zod.ZodDefault) {
        hasDefault = true;
        defaultValue = currentSchema._def.defaultValue();
        currentSchema = currentSchema._def.innerType;
      }
      if (currentSchema instanceof import_zod.ZodNullable) {
        currentSchema = currentSchema._def.innerType;
      }
    }
    if (currentSchema instanceof import_zod.ZodArray) {
      const arraySchema = currentSchema;
      const itemType = arraySchema._def.type;
      if (relationshipMappings && relationshipMappings[key]) {
        mongooseField.type = [import_mongoose.default.Schema.Types.ObjectId];
        mongooseField.ref = relationshipMappings[key];
      } else {
        if (itemType instanceof import_zod.ZodString) {
          mongooseField.type = [String];
        } else if (itemType instanceof import_zod.ZodNumber) {
          mongooseField.type = [Number];
        } else if (itemType instanceof import_zod.ZodBoolean) {
          mongooseField.type = [Boolean];
        } else if (itemType instanceof import_zod.ZodObject) {
          mongooseField.type = [import_mongoose.default.Schema.Types.Mixed];
        } else {
          mongooseField.type = [import_mongoose.default.Schema.Types.Mixed];
        }
      }
    } else if (relationshipMappings && relationshipMappings[key]) {
      mongooseField.type = import_mongoose.default.Schema.Types.ObjectId;
      mongooseField.ref = relationshipMappings[key];
    } else {
      switch (currentSchema.constructor) {
        case import_zod.ZodString:
          mongooseField.type = String;
          const stringSchema = currentSchema;
          stringSchema._def.checks.forEach((check) => {
            switch (check.kind) {
              case "min":
                mongooseField.minlength = check.value;
                break;
              case "max":
                mongooseField.maxlength = check.value;
                break;
              case "email":
                mongooseField.match = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
                break;
              case "regex":
                mongooseField.match = check.regex;
                break;
              case "includes":
                break;
              case "length":
                mongooseField.length = check.value;
                break;
              default:
                break;
            }
          });
          break;
        case import_zod.ZodNumber:
          mongooseField.type = Number;
          const numberSchema = currentSchema;
          numberSchema._def.checks.forEach((check) => {
            switch (check.kind) {
              case "min":
                mongooseField.min = check.value;
                break;
              case "max":
                mongooseField.max = check.value;
                break;
              case "int":
                mongooseField.validate = {
                  validator: Number.isInteger,
                  message: "{VALUE} is not an integer value"
                };
                break;
              default:
                break;
            }
          });
          break;
        case import_zod.ZodBoolean:
          mongooseField.type = Boolean;
          break;
        case import_zod.ZodDate:
          mongooseField.type = Date;
          break;
        case import_zod.ZodEnum:
          const enumSchema = currentSchema;
          mongooseField.type = String;
          mongooseField.enum = enumSchema.options;
          break;
        case import_zod.ZodLiteral:
          const literalSchema = currentSchema;
          switch (typeof literalSchema.value) {
            case "string":
              mongooseField.type = String;
              break;
            case "number":
              mongooseField.type = Number;
              break;
            case "boolean":
              mongooseField.type = Boolean;
              break;
            default:
              mongooseField.type = import_mongoose.default.Schema.Types.Mixed;
          }
          mongooseField.enum = [literalSchema.value];
          break;
        case import_zod.ZodObject:
          mongooseField.type = import_mongoose.default.Schema.Types.Mixed;
          break;
        default:
          mongooseField.type = import_mongoose.default.Schema.Types.Mixed;
      }
    }
    if (hasDefault) {
      mongooseField.default = defaultValue;
    }
    mongooseField.required = !isOptional;
    mongooseSchemaDefinition[key] = mongooseField;
  }
  return mongooseSchemaDefinition;
}
function createMongooseModel(modelName, zodSchema, relationshipMappings) {
  if (import_mongoose.default.models[modelName]) {
    return import_mongoose.default.models[modelName];
  }
  const mongooseSchemaDefinition = zodToMongoose(
    zodSchema,
    relationshipMappings
  );
  const schema = new import_mongoose.Schema(mongooseSchemaDefinition, { timestamps: true });
  return import_mongoose.default.model(modelName, schema);
}
function zodToObject(zodSchema, data, options) {
  const schemaToUse = options?.partial && zodSchema instanceof import_zod.ZodObject ? zodSchema.partial() : zodSchema;
  const parseResult = schemaToUse.safeParse(data);
  if (!parseResult.success) {
    const errorMessages = parseResult.error.errors.map((err) => `${err.path.join(".")} - ${err.message}`).join(", ");
    throw new Error(`Zod validation error: ${errorMessages}`);
  }
  return parseResult.data;
}

// src/zodToPrisma.ts
var import_zod2 = require("zod");
var enums = [];
function getEnumName(fieldName) {
  return `${capitalize(fieldName)}Enum`;
}
function capitalize(str) {
  if (!str) return str;
  return str.charAt(0).toUpperCase() + str.slice(1);
}
function camelCase(str) {
  if (!str) return str;
  return str.charAt(0).toLowerCase() + str.slice(1);
}
function zodTypeToPrisma(zodType, fieldName, schemaToModelName) {
  if (zodType instanceof import_zod2.ZodEffects) {
    return zodTypeToPrisma(zodType._def.schema, fieldName, schemaToModelName);
  }
  if (zodType instanceof import_zod2.ZodString) {
    return "String";
  }
  if (zodType instanceof import_zod2.ZodNumber) {
    if (zodType._def.checks?.some((check) => check.kind === "int")) {
      return "Int";
    }
    return "Float";
  }
  if (zodType instanceof import_zod2.ZodBoolean) {
    return "Boolean";
  }
  if (zodType instanceof import_zod2.ZodDate) {
    return "DateTime";
  }
  if (zodType instanceof import_zod2.ZodArray) {
    const elementType = zodType.element;
    const relatedModelName = schemaToModelName.get(elementType);
    if (relatedModelName) {
      return `${relatedModelName}[]`;
    }
    const prismaElementType = zodTypeToPrisma(
      elementType,
      fieldName,
      schemaToModelName
    );
    return `${prismaElementType}[]`;
  }
  if (zodType instanceof import_zod2.ZodEnum) {
    const enumName = getEnumName(fieldName);
    if (!enums.find((e) => e.name === enumName)) {
      enums.push({
        name: enumName,
        values: zodType._def.values
      });
    }
    return enumName;
  }
  if (zodType instanceof import_zod2.ZodNativeEnum) {
    const enumObj = zodType._def.values;
    const enumValues = Object.values(enumObj).filter(
      (v) => typeof v === "string"
    );
    const enumName = getEnumName(fieldName);
    if (!enums.find((e) => e.name === enumName)) {
      enums.push({
        name: enumName,
        values: enumValues
      });
    }
    return enumName;
  }
  if (zodType instanceof import_zod2.ZodLiteral) {
    const literalValue = zodType._def.value;
    if (typeof literalValue === "string") {
      return `String @default("${literalValue}")`;
    }
    if (typeof literalValue === "number") {
      return `Int @default(${literalValue})`;
    }
    if (typeof literalValue === "boolean") {
      return `Boolean @default(${literalValue})`;
    }
    throw new Error(`Unsupported ZodLiteral type: ${typeof literalValue}`);
  }
  if (zodType instanceof import_zod2.ZodObject) {
    const relatedModelName = schemaToModelName.get(zodType);
    if (relatedModelName) {
      return relatedModelName;
    }
    return "Json";
  }
  throw new Error(`Unsupported Zod type: ${zodType.constructor.name}`);
}
function extractPrismaFields(zodObject, schemaToModelName, modelName, modelNameToRelations, modelNameToSchema) {
  const shape = zodObject.shape;
  const fields = [];
  if (!shape.hasOwnProperty("id")) {
    fields.push({
      name: "id",
      type: "String",
      isRequired: true,
      isUnique: true,
      isId: true,
      isAutoIncrement: false,
      default: "uuid()"
    });
  }
  for (const [key, value] of Object.entries(shape)) {
    if (key === "id") {
      fields.push({
        name: "id",
        type: "String",
        isRequired: true,
        isId: true,
        isUnique: true,
        isAutoIncrement: false,
        default: "uuid()"
      });
      continue;
    }
    let isRequired = true;
    let isUnique = false;
    const isId = false;
    const isAutoIncrement = false;
    let defaultValue;
    let isEnum = false;
    let isRelation = false;
    let relationModel;
    let relationName;
    const relationFields = [];
    const references = [];
    let zodType = value;
    while (zodType instanceof import_zod2.ZodOptional || zodType instanceof import_zod2.ZodNullable || zodType instanceof import_zod2.ZodDefault || zodType instanceof import_zod2.ZodEffects) {
      if (zodType instanceof import_zod2.ZodOptional || zodType instanceof import_zod2.ZodNullable) {
        isRequired = false;
        zodType = zodType.unwrap();
      } else if (zodType instanceof import_zod2.ZodDefault) {
        const defaultVal = zodType._def.defaultValue();
        zodType = zodType._def.innerType;
        if (zodType instanceof import_zod2.ZodBoolean) {
          defaultValue = defaultVal ? "true" : "false";
        } else if (zodType instanceof import_zod2.ZodString) {
          defaultValue = `"${defaultVal}"`;
        } else if (zodType instanceof import_zod2.ZodNumber) {
          defaultValue = `${defaultVal}`;
        } else if (zodType instanceof import_zod2.ZodDate) {
          defaultValue = "now()";
        } else if (zodType instanceof import_zod2.ZodEnum || zodType instanceof import_zod2.ZodNativeEnum) {
          defaultValue = `${defaultVal}`;
        }
      } else if (zodType instanceof import_zod2.ZodEffects) {
        zodType = zodType._def.schema;
      }
    }
    if (key.toLowerCase().includes("email")) {
      isUnique = true;
    }
    if (zodType instanceof import_zod2.ZodEnum || zodType instanceof import_zod2.ZodNativeEnum) {
      isEnum = true;
    }
    if (zodType instanceof import_zod2.ZodObject || zodType instanceof import_zod2.ZodArray) {
      const elementType = zodType instanceof import_zod2.ZodArray ? zodType.element : zodType;
      if (elementType instanceof import_zod2.ZodObject) {
        const relatedModelName = schemaToModelName.get(elementType);
        if (relatedModelName) {
          isRelation = true;
          relationModel = relatedModelName;
          relationName = `${modelName}To${relatedModelName}`;
          const relatedRelations = modelNameToRelations.get(relatedModelName) || [];
          if (zodType instanceof import_zod2.ZodArray) {
            relatedRelations.push({
              relationName,
              relatedModel: modelName,
              type: "many",
              fieldName: key
            });
          } else {
            relatedRelations.push({
              relationName,
              relatedModel: modelName,
              type: "one",
              fieldName: key
            });
          }
          modelNameToRelations.set(relatedModelName, relatedRelations);
        }
      }
    }
    if (key.endsWith("Id") && (zodType instanceof import_zod2.ZodString || zodType instanceof import_zod2.ZodNumber)) {
      const relatedModelName = capitalize(key.slice(0, -2));
      if (modelNameToSchema.has(relatedModelName)) {
        fields.push({
          name: key,
          type: zodTypeToPrisma(zodType, key, schemaToModelName),
          isRequired,
          isUnique: true
          // Set to true for one-to-one relations
        });
        fields.push({
          name: key.slice(0, -2),
          type: relatedModelName,
          isRequired,
          isRelation: true,
          relationModel: relatedModelName,
          relationName: `${modelName}To${relatedModelName}`,
          relationFields: [key],
          references: ["id"]
        });
        const relatedRelations = modelNameToRelations.get(relatedModelName) || [];
        relatedRelations.push({
          relationName: `${modelName}To${relatedModelName}`,
          relatedModel: modelName,
          type: "one",
          fieldName: void 0
        });
        modelNameToRelations.set(relatedModelName, relatedRelations);
        continue;
      }
    }
    let prismaType;
    try {
      prismaType = zodTypeToPrisma(zodType, key, schemaToModelName);
    } catch (error) {
      console.error(`Error processing field "${key}": ${error.message}`);
      throw error;
    }
    fields.push({
      name: key,
      type: prismaType,
      isRequired,
      isUnique,
      isId,
      isAutoIncrement,
      default: defaultValue,
      isEnum,
      isRelation,
      relationModel,
      relationName,
      relationFields,
      references
    });
  }
  return fields;
}
function generatePrismaModel(modelName, zodSchema, schemaToModelName, modelNameToRelations, modelNameToSchema) {
  if (!(zodSchema instanceof import_zod2.ZodObject)) {
    throw new Error(
      "Only ZodObject schemas are supported for Prisma model generation."
    );
  }
  const fields = extractPrismaFields(
    zodSchema,
    schemaToModelName,
    modelName,
    modelNameToRelations,
    modelNameToSchema
  );
  const prismaFields = [];
  fields.forEach((field) => {
    let line = `  ${field.name} ${field.type}`;
    if (!field.isRequired && !field.isRelation) {
      line += "?";
    }
    if (field.isId) line += " @id";
    if (field.isUnique) line += " @unique";
    if (field.isAutoIncrement) line += " @default(autoincrement())";
    if (field.default && !field.isEnum) {
      line += ` @default(${field.default})`;
    }
    if (field.isEnum && field.default) {
      line += ` @default(${field.default})`;
    }
    if (field.isRelation && field.relationModel) {
      if (field.relationFields?.length) {
        line += ` @relation("${field.relationName}", fields: [${field.relationFields.join(
          ", "
        )}], references: [${field.references?.join(", ")}])`;
      } else if (field.type.endsWith("[]")) {
        line += ` @relation("${field.relationName}")`;
      } else {
        line += ` @relation("${field.relationName}")`;
      }
    }
    prismaFields.push(line);
  });
  const relations = modelNameToRelations.get(modelName) || [];
  relations.forEach((rel) => {
    const fieldName = rel.fieldName || camelCase(rel.relatedModel) + (rel.type === "many" ? "s" : "");
    const fieldType = `${rel.relatedModel}${rel.type === "many" ? "[]" : "?"}`;
    const line = `  ${fieldName} ${fieldType} @relation("${rel.relationName}")`;
    prismaFields.push(line);
  });
  return `model ${modelName} {
${prismaFields.join("\n")}
}`;
}
function generatePrismaEnums() {
  return enums.map((enumObj) => {
    const values = enumObj.values.map((v) => `  ${v}`).join("\n");
    return `enum ${enumObj.name} {
${values}
}`;
  }).join("\n\n");
}
function zodToPrisma(models, schemaToModelName, modelNameToSchema) {
  const modelNameToRelations = /* @__PURE__ */ new Map();
  const prismaModels = models.map(
    ({ name, schema }) => generatePrismaModel(
      name,
      schema,
      schemaToModelName,
      modelNameToRelations,
      modelNameToSchema
    )
  ).join("\n\n");
  const prismaEnums = generatePrismaEnums();
  const prismaSchemaHeader = `
datasource db {
  provider = "postgresql"
  url      = env("POSTGRESQL_DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}
`;
  return `${prismaSchemaHeader}

${prismaEnums}

${prismaModels}`;
}
function handleRelations(modelString, schema) {
  return modelString;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createMongooseModel,
  generatePrismaEnums,
  generatePrismaModel,
  handleRelations,
  zodToMongoose,
  zodToObject,
  zodToPrisma
});
